File: w1/w1d3/recursion.rb
require 'byebug'

def range(start_val, q)
  return [] if q <= start_val + 1
  arr = []
  arr << start_val + 1
  arr + range(start_val + 1, q)
end

def i_range(start_val, q)
  arr = []
  i = start_val + 1
  while i <
    arr << i
    i += 1
  end
  arr

end

def exp1(b, n)
  return 1 if n == 0

  b * exp1(b, n - 1)
end

def exp2(b, n)
  return 1 if n == 0
  return b if n == 1

  if n.even?
    base = exp2(b, n / 2)
    base * base
  else
    base = exp2(b, (n - 1) / 2)
    base * base * b
  end
end

class Array

  def deep_dup

    return self.dup if self.none? { |el| el.is_a?(Array) }

    copy = []
    self.each do |el|
      copy << el.deep_dup if el.is_a?(Array)
    end

    copy
  end

end

def fib(n)
  return [1] if n == 1
  return [1, 1] if n <= 2
  prev_num = fib(n - 1).last
  two_prev = fib(n - 2).last
  fib(n - 1) << prev_num + two_prev
end
=begin
def permutations(array)

  return [array] if array.length == 1

  result = []

  num = array.shift
  perms = permutations(array)


  perms.each do |perm|
    (0..perm.length).each do |idx|
      result << perm[0...idx] + [num] + perm[idx..-1]
    end
  end

  result

end
=end

def permutations(array)
  return array if array.length == 1
  perms = []
  array.each_index do |idx|
    # debugger
    shifted = array.rotate(idx)

    permutations(shifted[1..-1]).each do |arr|
      perms << [arr].unshift(shifted[0])
    end
  end
  perms.map(&:flatten)

end

def bsearch(array, target)
  return nil if array.empty?
  mid = array.length / 2

  if array[mid] == target
    return mid
  elsif array[mid] > target
    return bsearch(array[0...mid], target)
  else
    right = bsearch(array[mid + 1..-1], target)
    right.nil? ? nil : right + mid + 1
  end
end

def merge_sort(arr)
  return arr if arr.length <= 1
  mid = arr.length / 2
  left, right = arr[0..mid], arr[mid..-1]
  merge(merge_sort(left), merge_sort(right))
end

def merge(left, right)
  res = []
  until left.empty? || right.empty?
    if left[0] <= right[0]
      res << left.shift
    else
      res << right.shift
    end
  end
  res + left + right
end

def subsets(arr)
  return [[]] if arr.empty?
  subs = subsets(arr[0..-2])
  subs.concat(subs.map { |el| el += [arr.last] })
end

def greedy_make_change(amount, bank = [25, 10, 5, 1])
  return [amount] if bank.any? { |el| el == amount }
  change = []

  bank.each do |coin|
    if coin < amount
      change += [coin] + greedy_make_change(amount - coin, bank)
    end
    break if coin < amount
  end
  change
end

# 24, [10, 7, 1]

def make_better_change(amount, bank = [25, 10, 5, 1])
  coins = bank.select { |coin| coin <= amount }
  return nil if coins.empty?

  result = []
  coins.each do |coin|
    remainder = amount - coin
    if remainder > 0
      remainder_solution = make_better_change(remainder, coins)
      result << [coin] + remainder_solution unless remainder_solution.nil?
    else
      result << [coin]
    end
  end
  result.sort_by!(&:size).first
end

File: w5/w5d3/recursion.js
function range(start, end) {
  if (end < start) {
    return [];
  }
  const prevCase = range(start + 1, end);

  return [start] + prevCase;
}

// console.log(range(1, 5));

function sumRec(array) {
  if (array.length === 0) {
    return 0;
  }

  const prevCase = sumRec(array.slice(1));

  return array[0] + prevCase;
}

// console.log(sumRec([1,2,3]));

function sumIt(array) {
  let sum = 0;

  array.forEach(el => {
    sum += el;
  });

  return sum;
}

// console.log(sumIt([1,2,3]));

function expVer1(b, exp) {
  if (exp === 0) {
    return 1;
  } else {
    return b * expVer1(b, exp - 1);
  }
}

// console.log(expVer1(2, 4));

function expVer2(b, exp) {
  if (exp === 0) {
    return 1;
  } else if (exp === 1) {
    return b;
  } else {
    if (exp % 2 === 0) {
      return expVer2(b, exp / 2) * expVer2(b, exp / 2);
    } else {
      return b * (expVer2(b, (exp - 1) / 2)) * (expVer2(b, (exp - 1) / 2));
    }
  }
}

// console.log(expVer2(2, 3));

function fibRec(n) {
  switch (n) {
    case 0:
      return [];
    case 1:
      return [1];
    case 2:
      return [1, 1];
  }
  let pC = fibRec(n - 1);

  return pC.concat([pC[pC.length - 2] + pC[pC.length - 1]]);
}

// console.log(fibRec(5));


function bSearch(ary, tar) {
  if (ary.length === 0) {
    return null;
  }

  const mid = Math.floor(ary.length / 2);

  if (ary[mid] === tar) {
    return mid;
  } else if (ary[mid] > tar) {
    return bSearch(ary.slice(0, mid), tar);
  } else {
    const right = bSearch(ary.slice(mid + 1), tar);
    if (right === null) {
      return right;
    } else return right + mid + 1;
  }
}

// console.log(bSearch([1, 2, 3], 1));
// console.log(bSearch([1, 3, 4, 5, 9], 5));
// console.log(bSearch([1, 2, 3, 4, 5, 7], 6));

function makeChange(total, coins) {
  if (total === 0) {
    return [];
  }
  let bestChange = null;

  coins.forEach((coin, index) => {
    if (coin > total) {
      return;
    }

    const remain = total - coin;
    const rChange = makeChange(remain, coins.slice(index));

    if (!rChange) {
      return;
    }

    const change = [coin].concat(rChange);

    if (!bestChange || change.length < bestChange.length) {
      bestChange = change;
    }
  });

  return bestChange;
}

// console.log(makeChange(35, [10,7,1]));

function merge(left, right) {

  let result = [];

  while (left.length > 0 && right.length > 0) {
    if (left[0] > right[0]) {
      result.push(right.shift());
    } else {
      result.push(left.shift());
    }
  }

  return result.concat(left).concat(right);

}

function mergeSort(array) {
  if (array.length < 2) {
    return array;
  }

  const mid = Math.floor(array.length / 2);
  const left = mergeSort(array.slice(0, mid));
  const right = mergeSort(array.slice(mid));

  return merge(left, right);
}

// console.log(mergeSort([3,14,3,45,24,4,3,34]));

function subSets(array) {
  if (array.length < 1) {
    return [[]];
  }
  let lastArr = array[array.length - 1];
  let subs = subSets(array.slice(0,array.length-1));
  let result = subs.map(el => el.concat([lastArr]));
  return subs.concat(result);

}

// console.log(subSets([1, 2, 3]));
